utility/sourcemap_accessor.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/outputs/elasticsearch"
+)
+
+type SmapAccessor interface {
+	Fetch(smapId SmapID) (*sourcemap.Consumer, error)
+	RemoveFromCache(smapId SmapID)
+}
+
+type Enum string
+
+const (
+	InitError   Enum = "InitError"
+	AccessError Enum = "AccessError"
+	MapError    Enum = "MapError"
+	ParseError  Enum = "ParseError"
+)
+
+type SmapError struct {
+	Msg  string
+	Kind Enum
+}
+
+func (e SmapError) Error() string {
+	return fmt.Sprintf("%s", e.Msg)
+}
+
+type SmapID struct {
+	ServiceName    string
+	ServiceVersion string
+	Path           string
+}
+
+type SmapConfig struct {
+	CacheExpiration      time.Duration //seconds
+	CacheCleanupInterval time.Duration //seconds
+	ElasticsearchConfig  *common.Config
+	Index                string
+}
+
+type SourcemapAccessor struct {
+	esClients              []elasticsearch.Client
+	smapCache              *cache.Cache
+	cacheDefaultExpiration time.Duration
+	index                  string
+}
+
+func NewSourcemapAccessor(config SmapConfig) (*SourcemapAccessor, error) {
+	logp.Debug("sourcemap", "NewSourcemapAccessor created at Time.now %v for index", time.Now().Unix(), config.Index)
+	esClients, err := elasticsearch.NewElasticsearchClients(config.ElasticsearchConfig)
+	if err != nil || esClients == nil || len(esClients) == 0 {
+		err := SmapError{Msg: "Sourcemap ESClient cannot be initialized.", Kind: InitError}
+		logp.Err(err.Error())
+		return nil, err
+	}
+
+	smapCache := cache.New(config.CacheExpiration, config.CacheCleanupInterval)
+	return &SourcemapAccessor{
+		esClients:              esClients,
+		smapCache:              smapCache,
+		cacheDefaultExpiration: config.CacheExpiration,
+		index: fmt.Sprintf("%v*", config.Index),
+	}, nil
+}
+
+func (s *SourcemapAccessor) Fetch(smapId SmapID) (*sourcemap.Consumer, error) {
+	smapConsumer, err := s.fetchFromCache(smapId)
+	if err != nil {
+		logp.Err(fmt.Sprintf("Sourcemap fetching from Cache Error %s", err.Error()))
+	}
+	if smapConsumer == nil {
+		smapConsumer, err = s.fetchFromES(smapId)
+		if err != nil {
+			logp.Err(fmt.Sprintf("Sourcemap fetching from ES Error %s", err.Error()))
+			return nil, err
+		}
+		s.AddToCache(smapId, smapConsumer)
+	}
+	return smapConsumer, err
+}
+
+func (s *SourcemapAccessor) AddToCache(smapId SmapID, smap *sourcemap.Consumer) {
+	s.smapCache.Set(smapId.key(), smap, s.cacheDefaultExpiration)
+	logp.Debug("sourcemap", "Added smapId %v. Cache now has %v entries.", smapId.key(), s.smapCache.ItemCount())
+}
+
+func (s *SourcemapAccessor) RemoveFromCache(smapId SmapID) {
+	s.smapCache.Delete(smapId.key())
+	logp.Debug("sourcemap", "Removed smapId %v. Cache now has %v entries.", smapId.key(), s.smapCache.ItemCount())
+}
+
+func (s *SourcemapAccessor) fetchFromCache(smapId SmapID) (*sourcemap.Consumer, error) {
+	if s.smapCache == nil {
+		msg := "SourceMap Cache requested but not initialized"
+		return nil, SmapError{Msg: msg, Kind: InitError}
```

Comment:
`s.smapCache` can only be nil, if someone didn't use `NewSourcemapAccessor` to create their `SourcemapAccessor`. 

`AddToCache`  and `RemoveFromCache` already assumes `s.smapCache` is not nil.

I think we should also do that there. It will also save us some lines in `Fetch`